File: Python/Design/380. Insert Delete GetRandom O(1).py
class RandomizedSet:

    def __init__(self):
        self.elements_map = {}
        self.elements = []

    def insert(self, val: int) -> bool:
        if val in self.elements_map:
            return False
        self.elements_map[val] = len(self.elements)
        self.elements.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.elements_map:
            return False
        idx_to_remove = self.elements_map[val]
        last_num = self.elements[-1]

        self.elements_map[last_num] = idx_to_remove
        self.elements[idx_to_remove] = last_num
        self.elements = self.elements[:-1]
        self.elements_map.pop(val)
        return True

    def getRandom(self) -> int:
        return self.elements[int(random.random()*9999999 % len(self.elements))]

File: CPP/Design/380. Insert Delete GetRandom O(1).cpp
class RandomizedSet {
private:
    unordered_map<int, int> elements_map;
    vector<int> elements;
public:
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if (elements_map.find(val) != elements_map.end()) return false;
        elements_map[val] = elements.size();
        elements.push_back(val);
        return true;
    }
    
    bool remove(int val) {
        if (elements_map.find(val) == elements_map.end()) return false;
        int last_idx = elements.size()-1;
        int last_num = elements[last_idx];
        int index_to_remove = elements_map[val];
        elements_map[last_num] = index_to_remove;
        elements_map.erase(val);

        swap(elements[index_to_remove], elements[last_idx]);
        elements.pop_back();

        return true;
    }
    
    int getRandom() {
        int index = rand() % elements.size();
        return elements[index];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */